### ISSUE TITLE
Don't automatically truncate long index names

### ISSUE THREAD

**Issue by krainboltgreene** [tgriesser/knex#1233]
Postgres has a lexical limit of *n* characters for identifiers, which includes indexes:

> The system uses no more than NAMEDATALEN-1 bytes of an identifier; longer names can be written in commands, but they will be truncated. By default, NAMEDATALEN is 64 so the maximum identifier length is 63 bytes. If this limit is problematic, it can be raised by changing the NAMEDATALEN constant in src/include/pg_config_manual.h.

ActiveRecord, for example, throws and warns of long names. This avoids issues like this:

``` javascript
table.index(["name", "nationality", "ethnicity", "culture", "age", "sexuality", "gender", "height"])
table.index(["name", "nationality", "ethnicity", "culture", "age", "sexuality", "gender", "weight"])
```

``` sql
CREATE INDEX name_location_nationality_ethnicity_culture_age_sexuality_gender  ON profiles USING btree ("name", "nationality", "ethnicity", "culture", "age", "sexuality", "gender", "height")
CREATE INDEX name_location_nationality_ethnicity_culture_age_sexuality_gender  ON profiles USING btree ("name", "nationality", "ethnicity", "culture", "age", "sexuality", "gender", "weight")
```

Right now knex silently creates the truncated index.

**Comment by krainboltgreene** [comment]
I apologize in advance for the index name content if it bothers anyone, it was the best non-lorem ipsum example I could come up with that wouldn't sound made up. D:

**Comment by rhys-vdw** [comment]
Yes, this sounds good. I'm guessing it would need to be specified for each dialect?

**Comment by mercmobily** [comment]
Thanks for raising this important issue! You're absolutely right - the silent truncation of long index names is a problem that can lead to subtle bugs.

**Current Status**

This issue is still valid. Testing with current Knex confirms that generated index names can exceed database identifier limits:
- PostgreSQL: 63 bytes
- Oracle: 30 bytes (even more restrictive!)
- MySQL: 64 characters
- SQL Server: 128 characters

Your example still produces 75-character index names that PostgreSQL would silently truncate, potentially causing collisions.

**Why This Hasn't Been Fixed**

The main challenge is that any automatic solution would be a breaking change. Existing migrations rely on the current naming convention, and changing it could break database schemas in production.

**Current Workaround**

For now, provide explicit index names when dealing with multiple columns:
```javascript
table.index(
  ['name', 'nationality', 'ethnicity', 'culture', 'age', 'sexuality', 'gender', 'height'],
  'idx_demo_height'  // Explicit short name
);
```

**Proposed Path Forward**

I think we should implement a phased approach:

1. **v3.x** (non-breaking): Add warnings when generated names exceed database limits
2. **v4.0** (breaking): Implement smart truncation with hash suffixes to ensure uniqueness

For example, long names could become: `test_name_nationality_[8char_hash]_index`

This would prevent collisions while keeping names somewhat readable.

**Next Steps**

This needs an RFC to decide on the best approach. The solution should:
- Handle different database limits
- Ensure unique names
- Provide good developer experience
- Minimize breaking changes

I'm keeping this issue open as it represents a real problem that needs addressing.
